### DISCUSSION TITLE
Any way to enforce persisted IDs to be up to date?

### DISCUSSION THREAD

**Discussion by louy** [0no-co/gql.tada#310] [Q&A] [answered]
Hi,

I'm trying to enforce that persisted document IDs are always up to date, and I can see that gql.tada offers a code action to keep the id up to date with the document, but I can't see a way to enforce it to be up to date?

Is there a way to either:
- Make the code action a typescript error, so if the ID is not up to date our CI pipeline fails
- Programatically execute the code action, so we can check if there was a diff afterwards and fail if so
- Programatically extract a query document from a call site (with all its linked fragments) so we can generate our own persisted operations file. This is probably the least feasible

Appreciate any help/advice here

Best,

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

You should be getting a warning in your CI when running `pnpm gql-tada check`, you can make your CI fail when you enable `--fail-on-warn`. The relevant code is [here](https://github.com/0no-co/GraphQLSP/blob/main/packages/graphqlsp/src/diagnostics.ts#L270) we opt you into this code path when your hash starts with `sha256:`.

**Reply by louy** [reply]
Thanks!
It's a bit odd for me that some warnings fail that command but not others don't (e.g. `Unused co-located fragment definition(s)` and `Field(s) '*' are not used.`)

However, this does solve my problem, so thank you! Might be worth mentioning in the persisted operations documentation page, I'm happy to contribute if you agree

**Reply by louy** [reply]
Actually that command argument is just not working for me. Not ideal
``` console
$ npx gql-tada check -w
...
src/...
  6:46  warn    Unused co-located fragment definition(s) "..." in '...'

✖ 509 problems (0 errors, 509 warnings)
$ echo $?
0
```

However, the check command is sufficient. I can take the output from it and parse it, and fail when i detect a hash error

**Reply by louy** [reply]
If anyone is interested, I ended up creating a small script to read the output of the command above and do my own annotations/failure in github actions
```

      - run: npx gql-tada check > tada-output.txt 2>&1
      - uses: actions/github-script@v7
        with:
          script: |
            const contents = require('fs').readFileSync('./tada-output.txt', 'utf8');

            let currentFile;
            let foundErrors = false;

            for (const line of contents.split('\n')) {
              if (line.startsWith('⚠️')) {
                continue;
              }
              if (line.startsWith('::')) {
                // Workflow command
                continue;
              }
              if (!line.length) continue;

              if (!line.startsWith(' ')) {
                // File name line
                currentFile = line;
                continue;
              }

              const matches = line.match(/^\s+(\d+\:\d+)\s+([a-z]+)\s+(.*)$/);
              if (matches && matches.length === 4) {
                const [, loc, severity, message] = matches;

                const [startLine, startColumn] = loc.split(':').map(v => parseInt(v, 10));
                if (message.includes("persisted document's hash")) {
                  foundErrors = true;
                  core.error(message, { file: currentFile, startLine, startColumn });
                } else if (severity === 'warn') {
                  core.warning(message, { file: currentFile, startLine, startColumn });
                } else if (severity === 'error') {
                  foundErrors = true;
                  core.error(message, { file: currentFile, startLine, startColumn });
                }
              }
            }

            if (foundErrors) {
              process.exit(1);
            }
```
